pkg/infra/httpclient/httpclientprovider/http_client_provider.go

```diff
@@ -27,6 +26,7 @@ func New(cfg *setting.Cfg, tracer tracing.Tracer, features featuremgmt.FeatureTo
 		SetUserAgentMiddleware(userAgent),
 		sdkhttpclient.BasicAuthenticationMiddleware(),
 		sdkhttpclient.CustomHeadersMiddleware(),
+		AzureMiddleware(cfg),
```

Comment:
It's not optimal that every single outgoing datasource HTTP request will pass through the azure middleware which will look if the CustomOptions contains a certain key, i.e. no strong typing. For example [this](https://github.com/grafana/grafana/blob/2e6f14d17f832cdeddbbc4f63fa0f388f877a6d9/pkg/infra/httpclient/httpclientprovider/sigv4_middleware.go#L19-L21) makes things much cleaner and easier to reason about. My [comment](https://github.com/grafana/grafana-plugin-sdk-go/pull/468#pullrequestreview-897329203) in your other PR is highly relevant.

There's no feature toggle or setting to enable/disable azure auth since it's being used by builtin azure datasources, is that correctly understood?

Reply:
Not for every request, for every instantiation of `http.Client`, i.e. once per each datasource, given the client instance is preserved.

`AzureMiddleware` already has short return:
https://github.com/kostrse/grafana/blob/09da311c04ab69b0f94fe5fb51ef605adf291826/pkg/infra/httpclient/httpclientprovider/azure_middleware.go#L21
It excludes the middleware out of request/response round trip.

Once strongly-typed schema introduced to `httpclient.Options`, it will eliminate overhead of working with `CustomOptions`:

```go
func AzureMiddleware(cfg *setting.Cfg) httpclient.Middleware {
	return httpclient.NamedMiddlewareFunc(azureMiddlewareName, func(opts httpclient.Options, next http.RoundTripper) http.RoundTripper {
		if opts.Azure == nil || opts.Azure.Credentials == nil {
			return next
		}

		...

		return aztokenprovider.ApplyAuth(tokenProvider, opts.Azure.Scopes, next)
	})
}
```

I removed the feature toggle because, yes, the middleware will be used by Azure Monitor and Prometheus and it should work out of box, without any feature flags.

Since `infra/httpclient/httpclientprovider` is available only to internal plugins, external plugins will still need to manually register the middleware for themselves. Do we have a way to enable it by default for external plugins too?
Considering this, does it make sense to require all plugins (internal and external) to enable it explicitly for consistency?